File: data structure and algorithms/Math/587.erect-the-fence.go
package leetcode

//
// @lc app=leetcode id=587 lang=golang
//
// [587] Erect the Fence
//
// Notes
// Use the monotonic chain convex hull. Sorting points lexicographically, build
// lower and upper hulls while popping only clockwise turns; collinear boundary
// points remain on the hull. A set removes duplicate endpoints before returning
// coordinates. Time: O(n log n). Space: O(n).
//
// @lc code=start

import "sort"

type point587 struct {
	x int
	y int
}

func OuterTrees587(trees [][]int) [][]int {
	points := make([]point587, len(trees))
	for i, point := range trees {
		points[i] = point587{x: point[0], y: point[1]}
	}
	sort.Slice(points, func(i, j int) bool {
		if points[i].x != points[j].x {
			return points[i].x < points[j].x
		}
		return points[i].y < points[j].y
	})

	if len(points) <= 3 {
		answer := make([][]int, len(points))
		for i, point := range points {
			answer[i] = []int{point.x, point.y}
		}
		return answer
	}

	lower := []point587{}
	for _, point := range points {
		for len(lower) >= 2 && cross587(lower[len(lower)-2], lower[len(lower)-1], point) < 0 {
			lower = lower[:len(lower)-1]
		}
		lower = append(lower, point)
	}

	upper := []point587{}
	for i := len(points) - 1; i >= 0; i-- {
		point := points[i]
		for len(upper) >= 2 && cross587(upper[len(upper)-2], upper[len(upper)-1], point) < 0 {
			upper = upper[:len(upper)-1]
		}
		upper = append(upper, point)
	}

	boundary := map[point587]struct{}{}
	for _, point := range lower {
		boundary[point] = struct{}{}
	}
	for _, point := range upper {
		boundary[point] = struct{}{}
	}

	answer := make([][]int, 0, len(boundary))
	for point := range boundary {
		answer = append(answer, []int{point.x, point.y})
	}
	return answer
}

func cross587(a, b, c point587) int {
	return (b.x-a.x)*(c.y-a.y) - (b.y-a.y)*(c.x-a.x)
}

// @lc code=end

File: data structure and algorithms/Array/587.erect-the-fence.java
/*
 * @lc app=leetcode id=587 lang=java
 *
 * [587] Erect the Fence
 *
 * Andrew's monotonic chain convex hull. Sort points, build lower and upper
 * hulls, and pop only on clockwise turns so collinear boundary points stay on
 * the fence. Use a set to remove duplicate endpoints.
 *
 * Time: O(n log n). Space: O(n).
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

// @lc code=start
class Solution {
    public int[][] outerTrees(int[][] trees) {
        int[][] points = trees.clone();
        Arrays.sort(points, (a, b) -> a[0] == b[0] ? Integer.compare(a[1], b[1]) : Integer.compare(a[0], b[0]));
        if (points.length <= 3) {
            return points;
        }

        List<int[]> lower = new ArrayList<>();
        for (int[] point : points) {
            while (lower.size() >= 2 && cross(lower.get(lower.size() - 2), lower.get(lower.size() - 1), point) < 0) {
                lower.remove(lower.size() - 1);
            }
            lower.add(point);
        }

        List<int[]> upper = new ArrayList<>();
        for (int i = points.length - 1; i >= 0; i--) {
            int[] point = points[i];
            while (upper.size() >= 2 && cross(upper.get(upper.size() - 2), upper.get(upper.size() - 1), point) < 0) {
                upper.remove(upper.size() - 1);
            }
            upper.add(point);
        }

        Set<String> seen = new HashSet<>();
        List<int[]> answer = new ArrayList<>();
        for (int[] point : lower) {
            addPoint(answer, seen, point);
        }
        for (int[] point : upper) {
            addPoint(answer, seen, point);
        }
        return answer.toArray(new int[answer.size()][]);
    }

    private long cross(int[] a, int[] b, int[] c) {
        return (long) (b[0] - a[0]) * (c[1] - a[1])
                - (long) (b[1] - a[1]) * (c[0] - a[0]);
    }

    private void addPoint(List<int[]> answer, Set<String> seen, int[] point) {
        String key = point[0] + "#" + point[1];
        if (seen.add(key)) {
            answer.add(new int[] {point[0], point[1]});
        }
    }
}
// @lc code=end
